cli/internal/signals/signals.go

```diff
@@ -0,0 +1,60 @@
+package signals
+
+import (
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+)
+
+// Watcher watches for signals delivered to this process and provides
+// the opportunity for turbo to run cleanup
+type Watcher struct {
+	doneCh  chan struct{}
+	closed  bool
+	mu      sync.Mutex
+	closers []func()
+}
+
+// AddOnClose registers a cleanup handler to run when a signal is received
+func (w *Watcher) AddOnClose(closer func()) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.closers = append(w.closers, closer)
```

Comment:
`append` is not atomic, and _might_ reallocate, which is why you need to save the return value. 

This can race with getting a signal. If `AddOnClose` wins, your new close handler will run. If it loses, it won't.

Reply:
Thanks for the lesson!